### PR TITLE
Improve parsing of composite inline blocks

### LINKED ISSUE #51: '$'-delimited in-line math example fails in `syntax.md`
The $-delimited math [on line 352 of syntax.md](https://github.com/ExecutableBookProject/myst_parser/blob/develop/docs/using/syntax.md#L352) doesn't render properly at build.

Other instances of $-delimited inline math in the file seem to work though, not sure why this particular instance is not working.

N.B. switching to the myst-role syntax `{math}...` works fine.

### PR DESCRIPTION
**Pull request by chrisjsewell**
fixes #51, where for example a `$` within a code span was causing issues for later math blocks, i.e. for.

```
`$` $a=1$
``` 

Only a math block between the first and second dollar sign was being recognised as potential math (overriden in priority by the code block), since the math block regex was run once per paragraph.

Now the algorithm looks for the first span block match (code block or math block) and, if a block is found, records that block, advances the cursor position to the end of the block, and recomputes the next block.

### CODE REVIEW

**Comment by chrisjsewell** [comment]
@choldgraf since this is a change to the core parsing algorithm, before merging, we should read carefully through the CircleCI artefacts, to check for any unexpected behaviour not picked up by the unit tests.

**Comment by chrisjsewell** [comment]
Differences noted between 'Wealth Distribution Dynamics' examples:

- In RST you have `.. contents:: :depth: 2` (this relates to #60)
- In RST the whole header text is turned into a hyperlink, whereas in Myst it is only the trailing anchor.

**Comment by chrisjsewell** [comment]
Didn't see any issues in `syntax.md`, but I did note that RTD (right) doesn't appear to be rendering admonitions correctly, like from CircleCI artefact (left)?

image

**Comment by chrisjsewell** [comment]
`:math:numref:` is no working 😄 https://125-240151150-gh.circle-artifacts.com/0/html/using/syntax.html#equation-euler

**Comment by choldgraf** [comment]
that is quite strange that RTD has different behavior from Circle (circle looks fine to me FWIW). Perhaps that's something unrelated to this PR that we should figure out in a different issue? I wonder if RTD is loading (or not loading) some kind of asset that is different in CircleCI. Or, it could be that RTD is using a different version of Sphinx (if I remember it uses an older 1.X version)

**Comment by chrisjsewell** [comment]
> using a different version of Sphinx

That would be my first thought as well. Maybe you can force RTD to use a newer version of sphinx? I'm sure I've done something like this before, by using Conda.

**Comment by choldgraf** [comment]
yep, I think you can specify it in RTD.yml, I'll try it in another PR to see if that fixes it (may do some self-merging since we don't get the build results until it's merged)

**Comment by chrisjsewell** [comment]
> may do some self-merging since we don't get the build results until it's merged

You can add extra repo branches for RTD to build in the `Version` settings

**Comment by choldgraf** [comment]
hmmm, actually I take it back, our `[sphinx]` install option does specify 2.X, and the logs say it's running 2.4.2

**Comment by choldgraf** [comment]
OK I think it has to do with the Sphinx theme we're using, I opened up https://github.com/ExecutableBookProject/myst_parser/issues/68 to track it

**Comment by chrisjsewell** [comment]
Anyway, I'm happy to merge this PR now, if you are?

File: docs/using/syntax.md [hunk 1]
```diff
@@ -318,10 +318,7 @@ label: euler
 ---
 ```
 
-```{todo}
-Figure out why equation referencing didn't work
-```
-Euler's identity, equation {DOESN'T WORKeq}`euler`, was elected one of the
+Euler's identity, equation {math:numref}`euler`, was elected one of the
```
Comment (choldgraf) [inline docs/using/syntax.md]: yessss

File: tests/test_syntax/test_ast.py [hunk 2]
```diff
@@ -22,7 +22,17 @@ def ast_renderer():
...
-        ("issue_51", ["`$x_{hey}=it+is^{math}$` renders as $x_{hey}=it+is^{math}$."]),
+        ("multiline", ["$$a", "c", "b$$"]),
+        (
+            "issue_51",
+            [
+                "Math can be called in-line with single `$` characters around math.",
+                "For example, `$x_{hey}=it+is^{math}$` renders $x_{hey}=it+is^{math}$.",
+            ],
+        ),
+        ("in_link_content", ["[$a$](link)"]),
```
Comment (choldgraf) [inline tests/test_syntax/test_ast.py]: embedding equations inside links...this is some power-user stuff right here ;-)

**Pull request merged**